src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+
+        if nr == 0 {
+            break;
+        }
+
+        cnt += nr;
+    }
+    Ok(cnt)
+}
+
+pub fn write_all<W: Write>(f: &mut W, buf: &[u8]) -> Result<usize, Error> {
+    let mut cnt = 0;
+    while cnt < buf.len() {
+        let nw = f.write(&buf[cnt..buf.len()])
+            .map_err(Error::IOError)?;
+        cnt += nw;
+    }
+    Ok(cnt)
+}
+
+/// In-RAM trie storage.
+/// Used by TrieFileStorage to buffer the next trie being built.
+pub struct TrieRAM {
+    data: Vec<(TrieNodeType, TrieHash)>,
+    block_header: BlockHeaderHash,
+    readonly: bool,
+
+    read_count: u64,
+    read_backptr_count: u64,
+    read_node_count: u64,
+    read_leaf_count: u64,
+
+    write_count: u64,
+    write_node_count: u64,
+    write_leaf_count: u64,
+
+    total_bytes: usize
+}
+
+// Trie in RAM without the serialization overhead
+impl TrieRAM {
+    pub fn new(block_header: &BlockHeaderHash, capacity_hint: usize) -> TrieRAM {
+        TrieRAM {
+            data: Vec::with_capacity(capacity_hint),
+            block_header: block_header.clone(),
+            readonly: false,
+
+            read_count: 0,
+            read_backptr_count: 0,
+            read_node_count: 0,
+            read_leaf_count: 0,
+
+            write_count: 0,
+            write_node_count: 0,
+            write_leaf_count: 0,
+
+            total_bytes: 0
+        }
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn stats(&mut self) -> (u64, u64) {
+        let r = self.read_count;
+        let w = self.write_count;
+        self.read_count = 0;
+        self.write_count = 0;
+        (r, w)
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn node_stats(&mut self) -> (u64, u64, u64) {
+        let nr = self.read_node_count;
+        let br = self.read_backptr_count;
+        let nw = self.write_node_count;
+
+        self.read_node_count = 0;
+        self.read_backptr_count = 0;
+        self.write_node_count = 0;
+         
+        (nr, br, nw)
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn leaf_stats(&mut self) -> (u64, u64) {
+        let lr = self.read_leaf_count;
+        let lw = self.write_leaf_count;
+
+        self.read_leaf_count = 0;
+        self.write_leaf_count = 0;
+
+        (lr, lw)
+    }
+
+    /// Walk through the bufferred TrieNodes and dump them to f.
+    fn dump_traverse<F: Read + Write + Seek>(&mut self, f: &mut F, root: &TrieNodeType, hash: &TrieHash, parent_hash: &BlockHeaderHash) -> Result<u64, Error> {
+        let mut frontier : VecDeque<(TrieNodeType, TrieHash)> = VecDeque::new();
+
+        let mut node_data = vec![];
+        let mut offsets = vec![];
+
+        frontier.push_back((root.clone(), hash.clone()));
+
+        let mut ptr = BLOCK_HEADER_HASH_ENCODED_SIZE as u64;       // first 32 bytes is reserved for the parent block hash
+        
+        // step 1: write out each node in breadth-first order to get their ptr offsets
+        while frontier.len() > 0 {
```

Comment:
You can use `while let` here to simplify the loop:

```rust
while let Some(node, node_hash) = frontier.pop_front {
// calculate size
...
}
```

Reply:
Didn't know that.  Will update.

Reply:
Should be fixed now.